File: benchmarks/levenshtein/levenshtein.go
package main

import (
	"fmt"
	"strings"
)

func levenshtein(s, t string) int {
	m, n := len(s), len(t)
	if m == 0 {
		return n
	}
	if n == 0 {
		return m
	}

	prev := make([]int, n+1)
	curr := make([]int, n+1)
	for j := 0; j <= n; j++ {
		prev[j] = j
	}

	for i := 0; i < m; i++ {
		curr[0] = i + 1
		for j := 0; j < n; j++ {
			cost := 1
			if s[i] == t[j] {
				cost = 0
			}
			ins := curr[j] + 1
			del := prev[j+1] + 1
			sub := prev[j] + cost
			best := ins
			if del < best {
				best = del
			}
			if sub < best {
				best = sub
			}
			curr[j+1] = best
		}
		prev, curr = curr, prev
	}

	return prev[n]
}

func main() {
	s := strings.Repeat("the quick brown fox jumps over the lazy dog", 20)
	t := strings.Repeat("the slow brown fox leaps over the lazy cat", 20)
	fmt.Println(levenshtein(s, t))
}

File: benchmarks/levenshtein/levenshtein.f90
program levenshtein_bench
  implicit none
  character(len=:), allocatable :: s, t
  integer :: d

  s = repeat("the quick brown fox jumps over the lazy dog", 20)
  t = repeat("the slow brown fox leaps over the lazy cat", 20)
  d = levenshtein(s, t)
  print '(I0)', d

contains

  function levenshtein(s, t) result(dist)
    character(len=*), intent(in) :: s, t
    integer :: dist
    integer :: m, n, i, j, cost, ins, del, sub, best
    integer, allocatable :: prev(:), curr(:), tmp(:)

    m = len(s); n = len(t)
    if (m == 0) then
      dist = n; return
    end if
    if (n == 0) then
      dist = m; return
    end if

    allocate(prev(0:n), curr(0:n))
    do j = 0, n
      prev(j) = j
    end do

    do i = 1, m
      curr(0) = i
      do j = 1, n
        if (s(i:i) == t(j:j)) then
          cost = 0
        else
          cost = 1
        end if
        ins  = curr(j - 1) + 1
        del  = prev(j) + 1
        sub  = prev(j - 1) + cost
        best = min(ins, del, sub)
        curr(j) = best
      end do
      tmp = prev; prev = curr; curr = tmp
    end do

    dist = prev(n)
    deallocate(prev, curr)
  end function levenshtein

end program levenshtein_bench
